File: dart/lib/id57.dart
import 'dart:convert';
import 'dart:typed_data';

import 'package:blake3_dart/blake3_dart.dart';

const String alphabet =
    'ABCDEFGHJKLMNPQRSTUVWXYZabcdefghijkmnpqrstuvwxyz123456789';
const int defaultLength = 12;

const String Alphabet = alphabet;
const int DefaultLength = defaultLength;

const Set<int> _supportedLengths = <int>{2, 3, 4, 5, 6, 8, 10, 12, 16, 32, 57};
final Set<String> _alphabetSet = alphabet.split('').toSet();

abstract class Id57Exception implements Exception {
  const Id57Exception(this.message);

  final String message;

  @override
  String toString() => message;
}

class UnsupportedLengthException extends Id57Exception {
  const UnsupportedLengthException(this.length)
    : super(
        'id57: unsupported length; must be 2, 3, 4, 5, 6, 8, 10, 12, 16, 32, or 57; got $length',
      );

  final int length;
}

class LengthMismatchException extends Id57Exception {
  const LengthMismatchException(this.got, this.want)
    : super(
        'id57: value length does not match the requested length: got $got, want $want',
      );

  final int got;
  final int want;
}

class InvalidCharsetException extends Id57Exception {
  const InvalidCharsetException(this.character, this.index)
    : super(
        "id57: value contains a character outside the ID57 alphabet: character '$character' at index $index",
      );

  final String character;
  final int index;
}

bool isSupportedLength(int length) => _supportedLengths.contains(length);

bool IsSupportedLength(int length) => isSupportedLength(length);

void validateLength(int length) {
  if (!isSupportedLength(length)) {
    throw UnsupportedLengthException(length);
  }
}

void ValidateLength(int length) => validateLength(length);

BigInt _bytesToBigInt(List<int> bytes) {
  var number = BigInt.zero;
  for (final byte in bytes) {
    number = (number << 8) + BigInt.from(byte);
  }
  return number;
}

String fromDigest(List<int> digest, int length) {
  validateLength(length);

  var number = digest.isEmpty ? BigInt.zero : _bytesToBigInt(digest);
  final base = BigInt.from(alphabet.length);
  final buffer = List<String>.filled(length, 'A');

  for (var i = 0; i < length; i++) {
    final remainder = (number % base).toInt();
    buffer[i] = alphabet[remainder];
    number ~/= base;
  }

  return buffer.join();
}

String FromDigest(List<int> digest, int length) => fromDigest(digest, length);

String generate(List<int> input, int length) {
  validateLength(length);
  return fromDigest(blake3(Uint8List.fromList(input), 32), length);
}

String Generate(List<int> input, int length) => generate(input, length);

String generateString(String input, int length) {
  return generate(utf8.encode(input), length);
}

String GenerateString(String input, int length) =>
    generateString(input, length);

void validate(String value, int length) {
  validateLength(length);

  if (value.length != length) {
    throw LengthMismatchException(value.length, length);
  }

  for (var index = 0; index < value.length; index++) {
    final character = String.fromCharCode(value.codeUnitAt(index));
    if (!_alphabetSet.contains(character)) {
      throw InvalidCharsetException(character, index);
    }
  }
}

void Validate(String value, int length) => validate(value, length);

bool isValid(String value, int length) {
  try {
    validate(value, length);
    return true;
  } on Id57Exception {
    return false;
  }
}

bool IsValid(String value, int length) => isValid(value, length);

File: js/id57.js
import { blake3 } from '@noble/hashes/blake3';

/** AIS ID57 alphabet: Base62 minus 0, o, O, I, and l. */
export const Alphabet = 'ABCDEFGHJKLMNPQRSTUVWXYZabcdefghijkmnpqrstuvwxyz123456789';

/** Default ID57 output length. */
export const DefaultLength = 12;

const ALPHABET_SIZE = BigInt(Alphabet.length);
const SUPPORTED = new Set([2, 3, 4, 5, 6, 8, 10, 12, 16, 32, 57]);
const ALPHABET_SET = new Set(Alphabet);

/**
 * Reports whether length is one of the supported ID57 sizes.
 * @param {number} length
 * @returns {boolean}
 */
export function IsSupportedLength(length) {
  return SUPPORTED.has(length);
}

/**
 * Throws if length is not a supported ID57 size.
 * @param {number} length
 */
export function ValidateLength(length) {
  if (!IsSupportedLength(length)) {
    throw new Error(
      `id57: unsupported length; must be 2, 3, 4, 5, 6, 8, 10, 12, 16, 32, or 57; got ${length}`
    );
  }
}

/**
 * Encodes a BLAKE3 digest into a deterministic ID57 value.
 * A zero-length or all-zero digest produces all-'A' output.
 * @param {Uint8Array} digest
 * @param {number} length
 * @returns {string}
 */
export function FromDigest(digest, length) {
  ValidateLength(length);
  let n = digest.reduce((acc, b) => acc * 256n + BigInt(b), 0n);
  const buf = new Array(length);
  for (let i = 0; i < length; i++) {
    buf[i] = Alphabet[Number(n % ALPHABET_SIZE)];
    n = n / ALPHABET_SIZE;
  }
  return buf.join('');
}

/**
 * Hashes input with BLAKE3 and projects the digest into an ID57 value.
 * @param {Uint8Array} input
 * @param {number} length
 * @returns {string}
 */
export function Generate(input, length) {
  ValidateLength(length);
  return FromDigest(blake3(input), length);
}

/**
 * Convenience wrapper around Generate for UTF-8 string input.
 * @param {string} input
 * @param {number} length
 * @returns {string}
 */
export function GenerateString(input, length) {
  return Generate(new TextEncoder().encode(input), length);
}

/**
 * Checks that value has the correct length and only ID57 characters.
 * Throws a descriptive Error on the first violation.
 * @param {string} value
 * @param {number} length
 */
export function Validate(value, length) {
  ValidateLength(length);
  if (value.length !== length) {
    throw new Error(
      `id57: value length does not match the requested length: got ${value.length}, want ${length}`
    );
  }
  for (const ch of value) {
    if (!ALPHABET_SET.has(ch)) {
      throw new Error(
        `id57: value contains a character outside the ID57 alphabet: character '${ch}'`
      );
    }
  }
}

/**
 * Reports whether value is a valid ID57 value of the requested length.
 * @param {string} value
 * @param {number} length
 * @returns {boolean}
 */
export function IsValid(value, length) {
  try {
    Validate(value, length);
    return true;
  } catch {
    return false;
  }
}

export {
  IsSupportedLength as isSupportedLength,
  ValidateLength as validateLength,
  FromDigest as fromDigest,
  Generate as generate,
  GenerateString as generateString,
  Validate as validate,
  IsValid as isValid,
};